process.c

```diff
@@ -8563,10 +8584,12 @@ proc_warmup(VALUE _)
     RB_VM_LOCK_ENTER();
     rb_gc_prepare_heap();
     RB_VM_LOCK_LEAVE();
+
+    rb_objspace_each_objects(proc_warmup_each_object, NULL);
```

Comment:
This means we iterate the whole the heap a second time.

We could do this at the same time we go over objects to mark them as candidate to oldgen as to only loop once, but it means we'd potentially compute the coderange of objects that are about to be collected, so I'm not 100% sure which is most efficient.

Reply:
Hum, the issue is that I'd need to include more headers in `gc.c` to have access to `STR_NOFREE` etc.

Reply:
Ok, I tried that, what do you think now?

Reply:
We should test in CI if not checking `STR_NOFREE` is a problem. It shouldn't be a problem. I would prefer if we didn't handle `STR_NOFREE` as a special case.

Reply:
Well, if we compute the coderange on NOFREE strings potentially about to be swept, there is a risk the pointer is garbage.

That said, are NOFREE strings even public API?

Reply:
Yes, it can be created with `rb_str_new_static`.

Reply:
Hum, right. And the doc clearly state you should only pass it "string literals" meaning them being freed is no concern.

If an extension uses this for malloced memory, it's their fault.

Reply:
I'll remove the flag check and merge tomorrow.

Reply:
Good point, so I guess it should be fine to compute it for `STR_NOFREE` strings.